src/stochastic_swap.rs

```diff
@@ -0,0 +1,314 @@
+    let mut trial_layout = int_layout.clone();
+    let mut optimal_layout = int_layout.clone();
+
+    let num_gates: usize = gates.len() / 2;
+    let num_edges: usize = edges.len() / 2;
+
+    let mut need_copy;
+    let mut cost_reduced;
+    let mut depth_step: usize = 1;
+    let depth_max: usize = 2 * num_qubits + 1;
+    let mut min_cost: f64;
+    let mut new_cost: f64;
+    let mut dist: f64;
+
+    let mut optimal_start: usize = std::usize::MAX;
+    let mut optimal_end: usize = std::usize::MAX;
+    let mut optimal_start_qubit = std::usize::MAX;
+    let mut optimal_end_qubit = std::usize::MAX;
+
+    let mut scale = Array2::zeros((num_qubits, num_qubits));
+
+    let distribution = Normal::new(1.0, 1.0 / num_qubits as f64).unwrap();
+    let mut rng: Pcg64Mcg = Pcg64Mcg::seed_from_u64(seed);
+    let rand_arr: Vec<f64> = distribution
+        .sample_iter(&mut rng)
+        .take(num_qubits * (num_qubits + 1) / 2)
+        .collect();
+
+    compute_random_scaling(&mut scale, &cdist2, &rand_arr, num_qubits);
+
+    let input_qubit_set: HashSet<usize> = int_qubit_subset.iter().copied().collect();
+
+    while depth_step < depth_max {
+        let mut qubit_set = input_qubit_set.clone();
+        while !qubit_set.is_empty() {
+            min_cost = compute_cost(&scale.view(), &trial_layout, gates, num_gates);
+            // Try to decrease the objective function
+            cost_reduced = false;
+            // Loop over edges of coupling graph
+            need_copy = true;
+            for idx in 0..num_edges {
+                let start_edge = edges[2 * idx];
+                let end_edge = edges[2 * idx + 1];
+                let start_qubit = trial_layout.phys_to_logic[start_edge];
+                let end_qubit = trial_layout.phys_to_logic[end_edge];
+                if qubit_set.contains(&start_qubit) && qubit_set.contains(&end_qubit) {
+                    // Try this edge to reduce cost
+                    if need_copy {
+                        new_layout = trial_layout.clone();
+                        need_copy = false;
+                    }
+                    new_layout.swap(start_edge, end_edge);
+                    // compute objective function
+                    new_cost = compute_cost(&scale.view(), &new_layout, gates, num_gates);
+                    // record progress if we succeed
+                    if new_cost < min_cost {
+                        cost_reduced = true;
+                        min_cost = new_cost;
+                        optimal_layout = new_layout.clone();
+                        optimal_start = start_edge;
+                        optimal_end = end_edge;
+                        optimal_start_qubit = start_qubit;
+                        optimal_end_qubit = end_qubit;
+                        need_copy = true;
+                    } else {
+                        new_layout.swap(start_edge, end_edge);
+                    }
+                }
```

Comment:
I know it was just a mechanical port of the previous cython code and it's minor but I don't see a reason why `new_layout` variable is necessary.

```suggestion
                if qubit_set.contains(&start_qubit) && qubit_set.contains(&end_qubit) {
                    // Try this edge to reduce cost
                    trial_layout.swap(start_edge, end_edge);
                    // compute objective function
                    new_cost = compute_cost(&scale.view(), &trial_layout, gates, num_gates);
                    // record progress if we succeed
                    if new_cost < min_cost {
                        cost_reduced = true;
                        min_cost = new_cost;
                        optimal_layout = trial_layout.clone();
                        optimal_start = start_edge;
                        optimal_end = end_edge;
                        optimal_start_qubit = start_qubit;
                        optimal_end_qubit = end_qubit;
                    }
                    trial_layout.swap(start_edge, end_edge);
                }
```